### ISSUE TITLE
Working with Excessively Long Strings

### ISSUE THREAD

**Issue by benrbray** [gcanti/parser-ts#45]
Thanks for this neat library!  For fun I'm attempting to port Pandoc's markdown parser to `parser-ts`, but I'm noticing that the library really struggles with longer strings.  Each of the following examples results in an `Uncaught RangeError: Maximum call stack size exceeded` error.  The limit is even less if I start combining parsers.

```
run(S.many(C.space),     " ".repeat(1000));
run(S.many(C.alphanum),  "x".repeat(1000));
run(S.many(C.digit),     "9".repeat(1000));
run(S.many(C.char("x")), "x".repeat(1000));
```

Do you have any tips for working with longer strings?  Is this a fundamental limitation of this approach, or is there some hope of working around it by modifying the recursion primitives?

(related to #41)

```
$ npm list | grep parser-ts
1:markdown-parser-ts@ /home/benjamin/projects/markdown-parser
20:├── parser-ts@0.6.13
```

Thanks!

**Comment by IMax153** [comment]
Hey @benrbray! Thanks for letting us know! 

Unfortunately, I am unable to replicate the recursion issue. For context, here is a [CodeSandbox](https://codesandbox.io/s/parser-ts-recursion-demo-hzfcn?file=/src/index.ts) containing the code you posted above. If you check the console, you'll see that all four parsers run as expected.

In addition, we have regression tests in place for the fix for #41 that are currently passing:

https://github.com/gcanti/parser-ts/blob/4d58e69d031e02688805cc0056b4aa4136c8acea/test/string.ts#L30-L42

Could you perhaps provide more context?

**Comment by benrbray** [comment]
Thanks for the quick reply!  I think this will be slightly system- and browser-dependent.  I tested in a few different environments:

* Ubuntu / Chrome v87 -- stack overflow at string length 1000
* Ubuntu / Firefox v1.0 -- works at length 1000, stack overflow at 2000
* Windows 10 / Chrome v89 -- works at length 1000, stack overflow at 2000

Are you able to reproduce the error on strings of say, length `10000`?

As for solutions, I haven't dug too much into the `parser-ts` internals yet, but I wonder if it's possible to write a `many sat` function specialized to strings that avoids recursion by directly iterating over characters.  I'm not sure whether that sort of approach would be compatible with [`Stream`](https://github.com/gcanti/parser-ts/blob/master/src/Stream.ts), though.

**Comment by IMax153** [comment]
Hey @benrbray - I was in fact able to replicate it when I increased the string length to 10,000. However, I am not really sure where the errors are originating from. I will need to do some further work on this but don't have too much time in the near future so this one might take a bit.

**Comment by benrbray** [comment]
No worries!  I'll do some of my own digging and post any updates here.

**Comment by IMax153** [comment]
@benrbray - I tracked down the issue. Essentially, the issue stems from the `Parser`'s `many` combinator, which is used internally by the `many` combinator in the `string` module. This combinator has been notorious for causing issues with infinite recursion in the past.

I believe I have figured out a fix to the recursion issue which allows the `many` combinators to handle infinitely large iterations. Will be pushing a PR soon.
